include/boost/gil/image_processing/threshold.hpp

```diff
@@ -0,0 +1,156 @@
+//
+// Copyright 2019 Miral Shah <miralshah2211@gmail.com>
+//
+// Use, modification and distribution are subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+#ifndef BOOST_GIL_IMAGE_PROCESSING_THRESHOLD_HPP
+#define BOOST_GIL_IMAGE_PROCESSING_THRESHOLD_HPP
+
+#include <functional>
+#include <limits>
+#include <boost/gil/image.hpp>
+
+namespace boost { namespace gil {
+
+
+namespace detail {
+
+template
+<
+    typename SourceChannelT,
+    typename ResultChannelT,
+    typename SrcView,
+    typename DstView
+>
+void threshold_simple_impl
+(
+    SrcView const& src_view,
+    DstView const& dst_view,
+    std::function<ResultChannelT(SourceChannelT)> const& logic
```

Comment:
### Avoid std::function Part 1/3

I suggest to avoid use of `std::function` unless it is necessary.

Since `threshold_simple_impl` is internal function we have a full control over what is passed as `Operator` functor or lambda. Although it is a good idea and something nice to have, I think, it is not necessary to force the functor signature. So, passing lambda via the basic template parameter should be sufficient sufficient.

```
template
<
    typename SrcView,
    typename DstView,
    typename Operator
>
void threshold_simple_impl
(
    SrcView const& src_view,
    DstView const& dst_view,
    Operator const& threshold_op
)
```